tps-migration/src/modules/tps-permit/tps-permit.service.ts

```diff
@@ -137,88 +135,184 @@ export class TpsPermitService {
-              revision: tpsPermit.revision - 1,
-            },
-            {
-              documentId: document.documentId,
-            },
-          );
-
-          await this.tpsPermitRepository.delete({
-            migrationId: tpsPermit.migrationId,
-          });
+      return s3Object;
+    } catch (error) {
+      this.logger.log(
+        `Error while uploading permit# ${tpsPermit.permitNumber} docs to S3`,
+      );
+      this.logger.error(error);
+      await this.tpsPermitRepository.update(
+        { migrationId: tpsPermit.migrationId },
+        {
+          s3UploadStatus: S3uploadStatus.Error,
+          retryCount: tpsPermit.retryCount + 1,
+          lastUpdateTimestamp: new Date(),
+          lastUpdateUser: 'dbo',
+        },
+      );
+      throw new InternalServerErrorException(
+        'S3 Upload Failed for TPS Permit Number ' + tpsPermit.permitNumber,
+      );
+    }
+  }
+
+  @LogAsyncMethodExecution()
+  async createDocumentAndHandleErrors(
+    s3ObjectId: string,
+    s3Object: CompleteMultipartUploadCommandOutput,
+    tpsPermit: TpsPermit,
+    companyId: number,
+  ) {
+    try {
+      const document = await this.createDocument(
+        s3ObjectId,
+        s3Object,
+        tpsPermit,
+        companyId,
+      );
+      return document;
+    } catch (error) {
+      this.logger.log(
+        `Error while processing permit# ${tpsPermit.permitNumber}`,
+      );
+      this.logger.error(error);
+      throw new InternalServerErrorException(
+        `Error while processing permit# ${tpsPermit.permitNumber}`,
+      );
+    }
+  }
+
+  @LogAsyncMethodExecution()
+  async updatePermitWithDocumentId(tpsPermit: TpsPermit, documentId: string) {
+    await this.permitRepository.update(
+      {
+        tpsPermitNumber: tpsPermit.permitNumber,
+        revision: tpsPermit.revision - 1,
+      },
+      {
+        documentId: documentId,
+        lastUpdateTimestamp: new Date(),
+        lastUpdateUser: 'dbo',
+      },
+    );
+  }
+
+  @LogAsyncMethodExecution()
+  async deleteTpsPermit(tpsPermitId: number) {
+    await this.tpsPermitRepository.delete({
+      migrationId: tpsPermitId,
+    });
+  }
+
+  @LogAsyncMethodExecution()
+  handleUploadError(tpsPermit: TpsPermit, error: unknown) {
+    this.logger.log(`Error while processing permit# ${tpsPermit.permitNumber}`);
+    this.logger.error(error);
+  }
+
+  /**
+   * This method identifies stuck permits and verifies if their corresponding documents have been uploaded.
+   * If a document already exists, the process takes no further action.
+   * Otherwise, it attempts to upload the permit document again to S3 storage and updates the reference in the ORBC permit table.
+   */
+  @Cron(`${process.env.TPS_MONITORING_POLLING_INTERVAL || '0 0 1 * * *'}`)
+  @LogAsyncMethodExecution()
+  async processTpsStuckRecords() {
+    const today = new Date();
+    const yesterday = new Date(today);
+    yesterday.setDate(today.getDate() - 1);
```

Comment:
What will happen if today.getDate() == 1? Will this still work supplying 0 to setDate()?

Reply:
Alternately, if today.getDate() == 30 and prev month (Feb) has only 28 days

Reply:
day =1 sets the date to the last day of previous month

Reply:
second comment is dumb, first comment is handled automatically by JS by setting day to last day of prev month when it gets 0 passed in.